### PR TITLE
Add Mockprover::verify_par()

### PR DESCRIPTION
**Pull request by spherel**
k = 20, reduce verifying time from **1717.8s** to **55.2s**.

### CODE REVIEW

File: halo2_proofs/src/dev.rs [hunk 1]
```diff
@@ -670,6 +672,37 @@ impl<F: Field + Group> Assignment<F> for MockProver<F> {
     }
 }
 
+fn home_made_par_iter_flat_map<U: Send, Fun: Send + Sync, const ISPAR: bool>(
```
Comment (davidnevadoc) [inline halo2_proofs/src/dev.rs:675]: I suggest changing the name to something like: `parallelizable_iter_flat_map` or `custom_par_iter_flat_map`

File: halo2_proofs/src/dev.rs [hunk 2]
```diff
@@ -670,6 +672,37 @@ impl<F: Field + Group> Assignment<F> for MockProver<F> {
...
+        row_indexes
+            .par_iter()
+            .flat_map(|row| f(*row as i32))
+            .collect()
+    } else {
+        row_indexes.iter().flat_map(|row| f(*row as i32)).collect()
+    }
+}
+
+fn home_made_par_iter_map<U: Send + Default, Fun: Send + Sync, const ISPAR: bool>(
```
Comment (davidnevadoc) [inline halo2_proofs/src/dev.rs:692]: Same as above.

File: halo2_proofs/src/dev.rs [hunk 3]
```diff
@@ -1040,7 +1099,16 @@ impl<F: FieldExt> MockProver<F> {
                                 row,
                             })
                         }
-                    })
+                    };
+                    let indexes: Vec<usize> = (0..values.len()).into_iter().collect();
+                    let ret: Vec<VerifyFailure> = indexes
+                        .par_iter()
```
Comment (davidnevadoc) [inline halo2_proofs/src/dev.rs:1105]: Should this `par_iter()` depend on `ISPAR`? Maybe it needs its own `home_made_par_iter_filter_map` wrapper.

**Review by davidnevadoc — APPROVED** [review]
Pretty nice improvement! :)
Other than a couple of nits, everything looks good.

**Review by han0110 — APPROVED** [review]
LGTM!

**Pull request merged**